### PR TITLE
manifest envsubst

### PR DESCRIPTION
**Pull request by so-jelly**
**What this PR does / why we need it**:

This PR replaces variable references in manifests with environment variables.   It also adds the generated pet name to the environment so that it can be templated as well.

Fixes # https://github.com/kudobuilder/kuttl/issues/203

### CODE REVIEW

**Review by kensipe — CHANGES_REQUESTED** [review]
I'll take a deeper look later today... however I would move out of `test/utils`.   It was there prior to my getting involved but have always disliked it.   We don't use "utils" in Go generally.  the package name provides meaning and context to the use of a func or struct.   Could we please consider moving it to a more meaningful package?

We are not going to introduce a need copyright into our code base tho... so translate.go needs to change to a deps or be written.

**Comment by kensipe** [comment]
Thanks for the contribution...   We need a sign-off on the DCO to move forward

**Comment by gberche-orange** [comment]
thanks @so-jelly for this contrib! It would very much helm me with asserting cluster-wide resources dynamically instanciated by a kuttl and referencing the associated $NAMESPACE

Any chance for you to proceed with the DCO signature so that this feature can be merged,, such as running `git rebase --signoff HEAD~2` ?

**Comment by haarchri** [comment]
@so-jelly  can you rebase and add DCO - happy to see this change comes in ;)

**Comment by so-jelly** [comment]
I refactored to remove that licensed code. Some edge cases i considered is variables in GVK and keys.   I have thought through this a bit and I think it is sound.   I can add some test cases as well.

**Comment by so-jelly** [comment]
I created test cases for some additional field entries.  I am not sure what is up with Github review requests. tagging @kensipe in case the review request isn't sufficient.

**Review by jaypipes — APPROVED** [review]
We need this basically yesterday :) Any chance we can see this merged?

File: pkg/test/case.go [hunk 1]
```diff
@@ -386,6 +386,10 @@ func (t *Case) determineNamespace() *namespace {
 		ns.Name = fmt.Sprintf("kuttl-test-%s", petname.Generate(2, "-"))
 		ns.AutoCreated = true
 	}
+
+	if _, ok := os.LookupEnv("NAMESPACE"); ok { // user can supply this environment variable
```
Comment (jaypipes) [inline pkg/test/case.go:390]: should this be 
```suggestion
	if _, ok := os.LookupEnv("NAMESPACE"); !ok { // user can supply this environment variable
```
otherwise the user's existing NAMESPACE environment variable will be overridden...

**Review by jaypipes — CHANGES_REQUESTED** [review]

File: pkg/test/case.go (hunk 1 above)
Comment (so-jelly) [reply pkg/test/case.go]: lgtm

**Comment by kensipe** [comment]
@jaypipes I can provide details elsewhere... I'm just seeing this.   looks urgent and apologies for being out of pocket... for way too long.  I will review today/night and respond and/or merge

**Comment by dooman87** [comment]
Wonder if this is something that could be added. 

Our use case is we need a random string in the resource manifest so external resources are not clashing when running.

**Comment by porridge** [comment]
@dooman87 there is https://github.com/kudobuilder/kuttl/pull/387 that I have higher hopes for... but not enough time to work on it further at the moment.

As for random strings, have you investigated `generateName` ([docs](https://kubernetes.io/docs/reference/using-api/api-concepts/#generated-values))?

**Comment by dooman87** [comment]
Thanks @porridge, #387 looks promising! 

I looked into `generateName` but could you use it for any field? I have something like:

```yaml
apiVersion: ourcrd/v1beta1
kind: CustomObject
metadata:
  name: test-object
spec:
  path: /path/to/s3/need-random-value-here/
```

Because every test is running it its own namespace having the same name is fine, but we need different path for every test run.

I workarounded it by generating manifest and using `envsubst` but it does look hacky :)

**Comment by porridge** [comment]
No @dooman87 , it's specific to metadata.name I think.

Believe me, I know the pain of `envsubst` 😅
